internal/provider/kubernetes/gatewayclass.go

```diff
@@ -96,21 +102,27 @@ func (r *gatewayClassReconciler) Reconcile(ctx context.Context, request reconcil
 
 	var cc controlledClasses
 
+	found := false
 	for i := range gatewayClasses.Items {
 		if gatewayClasses.Items[i].Spec.ControllerName == r.controller {
 			cc.addMatch(&gatewayClasses.Items[i])
+			if gatewayClasses.Items[i].GetName() == request.Name {
+				found = true
+			}
 		}
 	}
+	if !found {
+		r.resourceTable.GatewayClasses.Delete(request.Name)
+	}
 
 	// no controlled gatewayclasses, trigger a delete
 	if len(cc.matchedClasses) == 0 {
 		r.log.Info("failed to find gatewayclass", "name", request.Name)
-		// TODO: Delete gatewayclass from the IR.
-		// xref: https://github.com/envoyproxy/gateway/issues/38
 		return reconcile.Result{}, nil
 	}
 
 	updater := func(gc *gwapiv1b1.GatewayClass, accepted bool) error {
+		r.resourceTable.GatewayClasses.Store(gc.GetName(), gc)
```

Comment:
Here I think we need to store the GC that is "Accepted: true", and remove all the other GCs with a matching controller that are "Accepted: false".

Note from past experience implementing this -- think about what happens if you have multiple GatewayClasses matching the controller string, and then the one that's currently "Accepted: true" gets deleted. Now we need to mark a new GatewayClass as "Accepted: true" (even though it itself has not changed or triggered a reconcile), and then also ensure all of the old GatewayClass's Gateways are removed from our internal model, and all of the Gateways for the new GatewayClass get added (even though the Gateways themselves may not have changed so are not triggering reconciles).